compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -63,6 +63,20 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 ExprKind::ForLoop(pat, head, body, opt_label) => {
                     return self.lower_expr_for(e, pat, head, body, *opt_label);
                 }
+                // Similarly, async blocks do not use `e.id` but rather `closure_node_id`.
+                ExprKind::Async(capture_clause, closure_node_id, block) => {
```

Comment:
Should we just remove the `closure_node_id` from the AST, and keep the previous logic?

Reply:
I tried that here: https://github.com/Swatinem/rust/commit/f62aa02c1885b574f61740e0b909d63cc8a08d78
But I was struggling with some HIR errors related to that. I might revisit that idea at some point.